packages/rps/src/core/app-data.ts

```diff
@@ -1,26 +1,73 @@
 import {Weapon} from './weapons';
+import {BigNumber, defaultAbiCoder, bigNumberify, keccak256} from 'ethers/utils';
+import {HashZero} from 'ethers/constants';
+import {randomHex} from '../utils/randomHex';
 
-export interface Resting {
-  type: 'resting';
+export enum PositionType {
+  Start, // 0
+  RoundProposed, // 1
+  RoundAccepted, // 2
+  Reveal, // 3
 }
+export interface RPSData {
+  positionType: PositionType;
+  stake: BigNumber; // uint256
+  preCommit: string; // bytes32
+  playerAWeapon: Weapon;
+  playerBWeapon: Weapon;
+  salt: string; // bytes32
+}
+export interface Start {
+  type: 'start';
+}
+
+export type RoundProposed = Pick<RPSData, 'stake' | 'preCommit'> & {type: 'roundProposed'};
+
+export type RoundAccepted = Pick<RPSData, 'stake' | 'preCommit'> & {type: 'roundAccepted'};
 
-export interface Propose {
-  type: 'propose';
-  stake: number;
-  preCommit: string;
+export type Reveal = Pick<RPSData, 'playerAWeapon' | 'playerBWeapon'> & {type: 'reveal'};
+
+export type AppData = Start | RoundProposed | RoundAccepted | Reveal;
+
+export function toRPSData(appData: AppData): RPSData {
```

Comment:
Would it be possible to remove this function from the external app-data api? It seems that, as a user of this code, I'd rather do:
```ts
const start = { type: 'start' };

const encoded = encodeAppData(start);
```
instead of
```ts
const start = { type: 'start' };

const encoded = encodeAppData(toRPSData(start));
```
Is there a reason why a user would ever need access to the `toRPSData` method?

Reply:
It's certainly possible and in fact I had it that way in an intermediate commit. 

I went this way because 
a) it enables the `decodeAppData` to be the inverse of `encodeAppData` :  I suppose it would be possible to do this even if the input type of `encodeAppData` was `appData` rather than `RPSData`, at the expense of quite a bit of branching logic to discriminate the union and return the correct thing.
b) I felt it was good to be quite explicit about when default values are being used for unspecified fields.

Reply:
Yeah, I would expect `decode` to return an `AppData` in this case too. You'll need to switch on the state type to accomplish this. That's worth doing, as we will need to do this in the app, and otherwise that responsibility is also put on the caller.

You'd probably still end up calling `toRPSData` in the `encodeAppData` function, so you'd still be explicit about the defaults at that point. It would just be hidden in the app code.

Reply:
👍